File: Data Structures/5. Swift/f. Graphs.swift
struct Edge<T> {
    var destination: GraphNode<T>
    var weight: Double?
}

class GraphNode<T: Hashable> {
    var value: T
    var edges: [Edge<T>]

    init(value: T) {
        self.value = value
        self.edges = []
    }
}

class Graph<T: Hashable> {
    private var nodes: [GraphNode<T>]

    init() {
        self.nodes = []
    }

    // MARK: - Node Operations

    /// Add a node to the graph.
    func addNode(value: T) {
        let newNode = GraphNode(value: value)
        nodes.append(newNode)
    }

    /// Remove a node from the graph.
    func removeNode(value: T) {
        nodes.removeAll { $0.value == value }
        for node in nodes {
            node.edges.removeAll { $0.destination.value == value }
        }
    }

    // MARK: - Edge Operations

    /// Add an edge between two nodes in the graph.
    func addEdge(sourceValue: T, destinationValue: T, weight: Double? = nil) {
        guard let sourceNode = getNode(value: sourceValue),
              let destinationNode = getNode(value: destinationValue) else {
            return
        }
        let edge = Edge(destination: destinationNode, weight: weight)
        sourceNode.edges.append(edge)
    }

    /// Remove an edge between two nodes in the graph.
    func removeEdge(sourceValue: T, destinationValue: T) {
        guard let sourceNode = getNode(value: sourceValue) else {
            return
        }
        sourceNode.edges.removeAll { $0.destination.value == destinationValue }
    }

    // MARK: - Search

    /// Depth-First Search traversal starting from a specific node value.
    func depthFirstSearch(startValue: T) -> [T] {
        guard let startNode = getNode(value: startValue) else {
            return []
        }
        var visited: Set<T> = []
        var result: [T] = []
        depthFirstSearchRec(node: startNode, visited: &visited, result: &result)
        return result
    }

    private func depthFirstSearchRec(node: GraphNode<T>, visited: inout Set<T>, result: inout [T]) {
        if visited.contains(node.value) {
            return
        }
        visited.insert(node.value)
        result.append(node.value)
        for edge in node.edges {
            depthFirstSearchRec(node: edge.destination, visited: &visited, result: &result)
        }
    }

    /// Breadth-First Search traversal starting from a specific node value.
    func breadthFirstSearch(startValue: T) -> [T] {
        guard let startNode = getNode(value: startValue) else {
            return []
        }
        var visited: Set<T> = []
        var result: [T] = []
        var queue: [GraphNode<T>] = [startNode]

        while !queue.isEmpty {
            let currentNode = queue.removeFirst()
            if !visited.contains(currentNode.value) {
                visited.insert(currentNode.value)
                result.append(currentNode.value)
                queue.append(contentsOf: currentNode.edges.map { $0.destination })
            }
        }

        return result
    }

    // MARK: - Utility

    private func getNode(value: T) -> GraphNode<T>? {
        return nodes.first { $0.value == value }
    }

    // MARK: - Check Graph Properties

    /// Check if the graph is cyclic (contains cycles).
    func isCyclic() -> Bool {
        var visited: Set<T> = []
        var stack: Set<T> = []

        for node in nodes {
            if isCyclicHelper(node: node, visited: &visited, stack: &stack) {
                return true
            }
        }
        return false
    }

    private func isCyclicHelper(node: GraphNode<T>, visited: inout Set<T>, stack: inout Set<T>) -> Bool {
        if stack.contains(node.value) {
            return true
        }

        if visited.contains(node.value) {
            return false
        }

        visited.insert(node.value)
        stack.insert(node.value)

        for edge in node.edges {
            if isCyclicHelper(node: edge.destination, visited: &visited, stack: &stack) {
                return true
            }
        }

        stack.remove(node.value)
        return false
    }

    // MARK: - Shortest Path

    /// Find the shortest path between two nodes using Dijkstra's algorithm.
    func shortestPath(from startValue: T, to endValue: T) -> [T] {
        guard let startNode = getNode(value: startValue),
              let endNode = getNode(value: endValue) else {
            return []
        }

        var distances: [T: Double] = [:] // Store distances from startNode
        var previousNodes: [T: T] = [:]  // Store the previous node in the shortest path
        var unvisited: Set<GraphNode<T>> = Set(nodes)
        distances[startValue] = 0

        while let currentNode = unvisited.min(by: { distances[$0.value]! < distances[$1.value]! }) {
            unvisited.remove(currentNode)

            for edge in currentNode.edges {
                let distance = (distances[currentNode.value] ?? .infinity) + (edge.weight ?? 0)
                if distance < (distances[edge.destination.value] ?? .infinity) {
                    distances[edge.destination.value] = distance
                    previousNodes[edge.destination.value] = currentNode.value
                }
            }
        }

        var shortestPath: [T] = []
        var current = endValue
        while current != startValue {
            guard let previous = previousNodes[current] else {
                return []  // No path exists
            }
            shortestPath.insert(current, at: 0)
            current = previous
        }
        shortestPath.insert(startValue, at: 0)
        return shortestPath
    }

    // ... Add more graph operations as needed ...
}

File: data_structures/1_cpp/f_graphs.cpp
#include <iostream>
#include <vector>
#include <unordered_set>
#include <unordered_map>
#include <queue>

template <typename T>
struct Edge {
    GraphNode<T>* destination;
    double weight;
};

template <typename T>
class GraphNode {
public:
    T value;
    std::vector<Edge<T>> edges;

    GraphNode(T val) : value(val) {}
};

template <typename T>
class Graph {
private:
    std::vector<GraphNode<T>*> nodes;

public:
    // MARK: - Node Operations

    /// Add a node to the graph.
    void addNode(T value) {
        GraphNode<T>* newNode = new GraphNode<T>(value);
        nodes.push_back(newNode);
    }

    /// Remove a node from the graph.
    void removeNode(T value) {
        nodes.erase(
            std::remove_if(nodes.begin(), nodes.end(),
                [value](const GraphNode<T>* node) { return node->value == value; }),
            nodes.end()
        );

        for (auto& node : nodes) {
            node->edges.erase(
                std::remove_if(node->edges.begin(), node->edges.end(),
                    [value](const Edge<T>& edge) { return edge.destination->value == value; }),
                node->edges.end()
            );
        }
    }

    // MARK: - Edge Operations

    /// Add an edge between two nodes in the graph.
    void addEdge(T sourceValue, T destinationValue, double weight = 0) {
        auto sourceNode = getNode(sourceValue);
        auto destinationNode = getNode(destinationValue);

        if (sourceNode && destinationNode) {
            Edge<T> edge = { destinationNode, weight };
            sourceNode->edges.push_back(edge);
        }
    }

    /// Remove an edge between two nodes in the graph.
    void removeEdge(T sourceValue, T destinationValue) {
        auto sourceNode = getNode(sourceValue);

        if (sourceNode) {
            sourceNode->edges.erase(
                std::remove_if(sourceNode->edges.begin(), sourceNode->edges.end(),
                    [destinationValue](const Edge<T>& edge) { return edge.destination->value == destinationValue; }),
                sourceNode->edges.end()
            );
        }
    }

    // MARK: - Search

    /// Depth-First Search traversal starting from a specific node value.
    std::vector<T> depthFirstSearch(T startValue) {
        std::unordered_set<T> visited;
        std::vector<T> result;
        auto startNode = getNode(startValue);

        if (startNode)
            depthFirstSearchRec(startNode, visited, result);

        return result;
    }

    void depthFirstSearchRec(GraphNode<T>* node, std::unordered_set<T>& visited, std::vector<T>& result) {
        if (visited.find(node->value) != visited.end())
            return;

        visited.insert(node->value);
        result.push_back(node->value);

        for (const auto& edge : node->edges)
            depthFirstSearchRec(edge.destination, visited, result);
    }

    /// Breadth-First Search traversal starting from a specific node value.
    std::vector<T> breadthFirstSearch(T startValue) {
        std::unordered_set<T> visited;
        std::vector<T> result;
        std::queue<GraphNode<T>*> queue;

        auto startNode = getNode(startValue);

        if (!startNode)
            return result;

        queue.push(startNode);
        visited.insert(startValue);

        while (!queue.empty()) {
            auto currentNode = queue.front();
            queue.pop();

            result.push_back(currentNode->value);

            for (const auto& edge : currentNode->edges) {
                if (visited.find(edge.destination->value) == visited.end()) {
                    visited.insert(edge.destination->value);
                    queue.push(edge.destination);
                }
            }
        }

        return result;
    }

    // MARK: - Utility

    GraphNode<T>* getNode(T value) {
        auto it = std::find_if(nodes.begin(), nodes.end(), [value](const GraphNode<T>* node) { return node->value == value; });
        return (it != nodes.end()) ? *it : nullptr;
    }

    // MARK: - Check Graph Properties

    /// Check if the graph is cyclic (contains cycles).
    bool isCyclic() {
        std::unordered_set<T> visited;
        std::unordered_set<T> stack;

        for (const auto& node : nodes) {
            if (isCyclicHelper(node, visited, stack))
                return true;
        }

        return false;
    }

    bool isCyclicHelper(GraphNode<T>* node, std::unordered_set<T>& visited, std::unordered_set<T>& stack) {
        if (stack.find(node->value) != stack.end())
            return true;

        if (visited.find(node->value) != visited.end())
            return false;

        visited.insert(node->value);
        stack.insert(node->value);

        for (const auto& edge : node->edges) {
            if (isCyclicHelper(edge.destination, visited, stack))
                return true;
        }

        stack.erase(node->value);
        return false;
    }

    // MARK: - Shortest Path

    /// Find the shortest path between two nodes using Dijkstra's algorithm.
    std::vector<T> shortestPath(T startValue, T endValue) {
        std::unordered_map<T, double> distances; // Store distances from startNode
        std::unordered_map<T, T> previousNodes;   // Store the previous node in the shortest path
        std::unordered_set<GraphNode<T>*> unvisited;
        std::vector<T> shortestPath;

        for (auto& node : nodes) {
            unvisited.insert(node);
            distances[node->value] = std::numeric_limits<double>::infinity();
        }

        distances[startValue] = 0;

        while (!unvisited.empty()) {
            auto currentNode = *std::min_element(unvisited.begin(), unvisited.end(),
                [&distances](GraphNode<T>* a, GraphNode<T>* b) { return distances[a->value] < distances[b->value]; });

            unvisited.erase(currentNode);

            for (const auto& edge : currentNode->edges) {
                double distance = distances[currentNode->value] + edge.weight;
                if (distance < distances[edge.destination->value]) {
                    distances[edge.destination->value] = distance;
                    previousNodes[edge.destination->value] = currentNode->value;
                }
            }
        }

        auto current = endValue;
        while (current != startValue) {
            auto it = previousNodes.find(current);
            if (it == previousNodes.end())
                return {}; // No path exists

            shortestPath.insert(shortestPath.begin(), current);
            current = it->second;
        }

        shortestPath.insert(shortestPath.begin(), startValue);
        return shortestPath;
    }

    // ... Add more graph operations as needed ...
};

int main() {
    Graph<int> graph;

    graph.addNode(1);
    graph.addNode(2);
    graph.addNode(3);

    graph.addEdge(1, 2);
    graph.addEdge(2, 3);
    graph.addEdge(3, 1);

    std::cout << "Is cyclic: " << (graph.isCyclic() ? "true" : "false") << std::endl;

    std::vector<int> dfsResult = graph.depthFirstSearch(1);
    std::cout << "DFS result: ";
    for (int val : dfsResult)
        std::cout << val << " ";
    std::cout << std::endl;

    std::vector<int> bfsResult = graph.breadthFirstSearch(1);
    std::cout << "BFS result: ";
    for (int val : bfsResult)
        std::cout << val << " ";
    std::cout << std::endl;

    std::vector<int> shortestPathResult = graph.shortestPath(1, 3);
    std::cout << "Shortest path from 1 to 3: ";
    for (int val : shortestPathResult)
        std::cout << val << " ";
    std::cout << std::endl;

    return 0;
}
